### PR TITLE
`MidCircuitMeasure` inherits `MidMeasureMP`

### PR DESCRIPTION
**Pull request by mudit2812**
**Context:**
`MidCircuitMeasure` inherits `HybridOp` but only relies on its tracing functionality and nothing else, but also contains a lot of attributes that make it more appropriate for it to be a child of `MidMeasureMP` instead of `HybridOp`.

**Description of the Change:**
* `MidCircuitMeasure` inherits `MidMeasureMP` and uses useful attributes for the new parent class.
* Update `MidMeasureMP.trace_quantum` to be compatible with the above change.

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

**Comment by mudit2812** [comment]
Tests won't pass until https://github.com/PennyLaneAI/pennylane/pull/5888 is merged.

File: frontend/catalyst/api_extensions/quantum_operators.py [hunk 1]
```diff
@@ -16,8 +16,8 @@
 This module contains public API functions for quantum operators which are not
 included in PennyLane, or whose behaviour needs to be adapted for Catalyst.
 """
-
```
[the remark this answers is not in the document]
Comment (mudit2812) [reply frontend/catalyst/api_extensions/quantum_operators.py:19]: It make be `black` 😅

File: frontend/catalyst/api_extensions/quantum_operators.py [hunk 2]
```diff
@@ -147,17 +153,14 @@ def circuit():
...
     if postselect is not None and postselect not in [0, 1]:
         raise TypeError(f"postselect must be '0' or '1', got {postselect}")
 
     m = new_inner_tracer(ctx.trace, get_aval(True))
     MidCircuitMeasure(
-        in_classical_tracers=in_classical_tracers,
-        out_classical_tracers=[m],
-        regions=[],
+        # Copy, so wires remain unmodified
+        wires=wires.copy(),
```
[the remark this answers is not in the document]
Comment (mudit2812) [reply frontend/catalyst/api_extensions/quantum_operators.py:162]: Not particularly. Now that I also think about it, we don't need it since the wires get cast inside a `Wires` object now so we wouldn't be able to mutate them anyway.

File: frontend/catalyst/api_extensions/quantum_operators.py [hunk 3]
```diff
@@ -312,49 +315,41 @@ def cond_fn():
...
+        assert len(eqn.outvars[:-1]) == 1, f"{eqn.outvars=}\n{self.mv=}"
+
+        if trace.getvar(self.mv) not in set(
+            [
+                *sum([e.outvars for e in ctx.frames[trace].eqns[:-1]], []),
+                *ctx.frames[trace].invars,
+                *ctx.frames[trace].constvar_to_val.keys(),
+            ]
+        ):
+            eqn.outvars[0] = trace.getvar(self.mv)
```
[the remark this answers is not in the document]
Comment (mudit2812) [reply frontend/catalyst/api_extensions/quantum_operators.py:345]: To me it looks like it's making `mv` the output of the current line in the jaxpr if it's not already present anywhere else in the trace. Admittedly I don't fully understand the logic or the jargon, but I get errors without this block.

File: doc/changelog.md [hunk 4]
```diff
@@ -5,9 +5,13 @@
 * `qjit` adheres to user-specified `mcm_method` given to the `QNode`.
   [(#798)](https://github.com/PennyLaneAI/catalyst/pull/798)
 
+* `MidCiruitMeasure` now inherits `MidMeasureMP` instead of `HybridOp`, allowing much better integration with PennyLane for
+  workflows with mid-circuit measurements.
+  [(#832)](https://github.com/PennyLaneAI/catalyst/pull/832)
```
[the remark this answers is not in the document]
Comment (mudit2812) [reply doc/changelog.md:10]: I didn't even realize Catalyst had an "Internal changes" section in the changelog. Pennylane doesn't have one 🤔

File: frontend/catalyst/api_extensions/quantum_operators.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (dime10) [reply frontend/catalyst/api_extensions/quantum_operators.py:345]: @grwlf Can you add a comment about what this block does?

File: doc/changelog.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (dime10) [reply doc/changelog.md:10]: Yeah we've added it because there are things that are "user" targeted, and things that are "developer" targeted. The internal changes would mainly affect people poking around our package but not users.

File: frontend/catalyst/api_extensions/quantum_operators.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (grwlf) [reply frontend/catalyst/api_extensions/quantum_operators.py:345]: Hmm, looks like it is an approximate copy of the code residing in the `bind_overwrite_classical_tracers` [here](https://github.com/PennyLaneAI/catalyst/blob/aa19452e2c5bcf8f6aa0ee4a226daf47ff28f7b7/frontend/catalyst/jax_tracer.py#L464). This code handles a rather weird situation happening when we return dynamic dimensions from loops, so let me explain that.

Consider the following "identity-loop" program:

``` python
@qjit
@qnode(...)
def circuit(dim:int):
   @for_loop(0,10,1)
   def loop(i, a):
      return a
   b = loop(jnp.zeros([dim], float))
   ...
```

In order to trace this loop, Catalyst would parse `a` arguments into `in_type`, create inner-scope tracers, get the results, deduce `out_type` containing a reference to the input dimension and finally create an outer scope tracers for `b`. Following the InDBIdx/OutDBIdx semantics, it will re-use the `dim` tracer for the shape of `b`, which means that the expanded list of classical loop results would look something like `dim:i64[] b:f64[dim]`. If we directly interpret these results we would get a double-assignment for `dim`, e.g.

```
let dim:i64[] in # <--- First assignment
a:f64[dim] = jnp.zeros([dim], float)
dim:i64[]  b:f64[dim] =  loop({let dim:i64[] i:i64[] a:f64[dim]  in  (dim, a) }) dim a
^^^
Second assignment
```

The mentioned code avoids this by skipping the already-defined variables. We can always do this because we know that in the original program we used the same tracer in all cases.

[the remark this answers is not in the document]
Comment (grwlf) [reply frontend/catalyst/api_extensions/quantum_operators.py:345]: Speaking of this MCM code snippet, I would say that the `if var not in set(...)` condition should always fire, so my suggestion is to replace it with just
`eqn.outvars[0] = trace.getvar(self.mv)`
or, better, try calling `bind_overwrite_classical_tracers( ..., out_expanded_tracers = [self.mv], ... )`

[the remark this answers is not in the document]
Comment (dime10) [reply frontend/catalyst/api_extensions/quantum_operators.py:345]: > Speaking of this MCM code snippet, I would say that the if var not in set(...) condition should always fire, so my suggestion is to replace it with just eqn.outvars[0] = trace.getvar(self.mv)

Nice, even better 👍 

> try calling bind_overwrite_classical_tracers( ..., out_expanded_tracers = [self.mv], ... )
>
> We might need to move it out of this HybridOp class which actually was my initial proposal.

Yes we didn't want the MCM to be a HybridOp anymore, and in case we've been wanting to move the HybridOp class out of `jax_tracer` so making the `bind_overwrite_classical_tracers` standalone might be a good idea. We'll take a look next quarter :)

[the remark this answers is not in the document]
Comment (grwlf) [reply frontend/catalyst/api_extensions/quantum_operators.py:345]: Another suggestion: can we actually use both HybridOp and MidMeasureMP as base classes for MidCircuitMeasure, rather than changing from one to another?

[the remark this answers is not in the document]
Comment (dime10) [reply frontend/catalyst/api_extensions/quantum_operators.py:345]: > Another suggestion: can we actually use both HybridOp and MidMeasureMP as base classes for MidCircuitMeasure, rather than changing from one to another?

My preference was not to because it doesn't have a body, and reserve HybridOp for operations with nested scopes.

**Comment by dime10** [comment]
@mudit2812 Any chance we will see this improvement land or is it considered abandoned? 😅

**Comment by mudit2812** [comment]
@dime10 I wouldn't call it abandoned. Definitely think it's possible to get this in for 0.8, but it's just not a priority at the moment.
